ipn/ipnlocal/network-lock.go

```diff
@@ -122,6 +136,67 @@ func (b *LocalBackend) tkaFilterNetmapLocked(nm *netmap.NetworkMap) {
 	}
 }
 
+// rotationTracker determines the set of node keys that are made obsolete by key
+// rotation.
+//   - for each SigRotation signature, all previous node keys referenced by the
+//     nested signatures are marked as obsolete.
+//   - if there are multiple SigRotation signatures tracing back to the same
+//     wrapping pubkey (e.g. if a node is cloned with all its keys), we keep
+//     just one of them, marking the others as obsolete.
+type rotationTracker struct {
+	// obsolete is the set of node keys that are obsolete due to key rotation.
+	obsolete set.Set[key.NodePublic]
+
+	// byWrappingKey keeps track of rotation details per wrapping pubkey.
+	byWrappingKey map[string][]sigRotationDetails
+}
+
+// sigRotationDetails holds information about a node key signed by a SigRotation.
+type sigRotationDetails struct {
+	np          key.NodePublic
+	numPrevKeys int
+}
+
+// addRotationDetails records the rotation signature details for a node key.
+func (r *rotationTracker) addRotationDetails(np key.NodePublic, d *tka.RotationDetails) {
+	r.obsolete.Make()
+	r.obsolete.AddSlice(d.PrevNodeKeys)
+	rd := sigRotationDetails{
+		np:          np,
+		numPrevKeys: len(d.PrevNodeKeys),
+	}
+	if r.byWrappingKey == nil {
+		r.byWrappingKey = make(map[string][]sigRotationDetails)
+	}
+	wp := string(d.WrappingPubkey)
+	r.byWrappingKey[wp] = append(r.byWrappingKey[wp], rd)
+}
+
+// obsoleteKeys returns the set of node keys that are obsolete due to key rotation.
+func (r *rotationTracker) obsoleteKeys() set.Set[key.NodePublic] {
+	for _, v := range r.byWrappingKey {
+		// Sort by decreasing number of previous keys.
+		sort.SliceStable(v, func(i, j int) bool {
+			return v[i].numPrevKeys > v[j].numPrevKeys
+		})
+		skipped := false
+		for _, rd := range v {
+			if r.obsolete.Contains(rd.np) {
+				continue
+			}
+			// Reached the node that has the largest number of previous keys in
+			// its signature. Skip it, and add the rest to the list of obsolete
+			// keys.
```

Comment:
what happens if there are multiple nodes with the same numPrevKeys?
like a forked rotation chain, with 2 "heads".

Reply:
We do indeed take the first one, which I believe should be deterministic: netmap.Peers is sorted by Node.ID, and we use stable sorting to determine the "latest" one. I've added a comment describing this.

Not sure about obsoleting all keys if we have multiple candidates. Perhaps the implicit contract here is that a WrappingPubkey is allowed to sign a single node key, and while the goal of the rotation tracker is to ensure that maximum of 1, adding heuristics that will filter out all keys might have some unintended consequences? But also I can't imagine a practical case like that in a well behaved deployment, so happy to do that if you think it's safer.

Reply:
Yes, let's be paranoid here and reject all equally-long chains. This situation might (?) happen due to node cloning. So maybe let's add a log here too, to make it easier to diagnose.

Reply:
We can also make this a non-issue in practice by making sure control purges signatures from its database when a rotation occurs. Control should prevent node key rotation chain forking because when a rotation occurs it invalidates the old node key. Each machine key is only permitted one active node key per user identity, and a node key rotation doesn't change user identity. Maybe there's an edge case with tags.

Reply:
We do not trust control for this feature, the client should be as defensive as possible IMO

Reply:
Sorry, I framed that poorly.

I mean that we should ensure that control never allows this situation to exist. By doing that, it's okay if the client has a not-great UX in this situation, as long as it does something sensibly secure.

Reply:
I have extended this to filter out all keys signed by the same wrapping pubkey if there are multiple valid signatures with equally long rotation chains. PTAL?